s2spy/preprocess.py

```diff
@@ -0,0 +1,217 @@
+from typing import Tuple
+from typing import Union
+import numpy as np
+import scipy.stats
+import xarray as xr
+
+
+def _linregress(x: np.ndarray, y: np.ndarray) -> Tuple[float, float]:
+    """Calculate the slope and intercept between two arrays using scipy's linregress.
+
+    Used to make linregress more ufunc-friendly.
+
+    Args:
+        x: First array.
+        y: Second array.
+
+    Returns:
+        slope, intercept
+    """
+    slope, intercept, _, _, _ = scipy.stats.linregress(x, y)
+    return slope, intercept
+
+
+def _trend_linear(data: Union[xr.DataArray, xr.Dataset]) -> dict:
+    """Calculate the linear trend over time.
+
+    Args:
+        data: The input data of which you want to know the trend.
+
+    Returns:
+        Dictionary containing the linear trend information (slope and intercept)
+    """
+    slope, intercept = xr.apply_ufunc(
+        _linregress,
+        data["time"].astype(float),
+        data,
+        input_core_dims=[["time"], ["time"]],
+        output_core_dims=[[], []],
+        vectorize=True,
+    )
+    return {"slope": slope, "intercept": intercept}
+
+
+def _apply_linear_trend(data: Union[xr.DataArray, xr.Dataset], trend: dict):
+    """Apply a previously calclulated linear trend to (new) data."""
+    return data - trend["intercept"] - trend["slope"] * (data["time"].astype(float))
+
+
+def _get_trend(data: Union[xr.DataArray, xr.Dataset], method: str):
+    """Calculate the trend, with a certain method. Only linear is implemented."""
+    if method == "linear":
+        return _trend_linear(data)
+    raise ValueError(f"Unkown detrending method '{method}'")
+
+
+def _apply_trend(data: Union[xr.DataArray, xr.Dataset], method: str, trend: dict):
+    """Apply a previously calculated trend to (new) data. Only linear is implemented."""
+    if method == "linear":
+        return _apply_linear_trend(data, trend)
+    raise NotImplementedError
+
+
+def _get_climatology(data: Union[xr.Dataset, xr.DataArray]):
+    """Calculate the climatology of timeseries data."""
+    return data.groupby("time.dayofyear").mean("time")
+
+
+def _check_input_data(data: Union[xr.DataArray, xr.Dataset]):
+    """Check the input data for compatiblity with the preprocessor.
+
+    Args:
+        data: Data to validate.
+
+    Raises:
+        ValueError: If the input data is of the wrong type.
+        ValueError: If the input data does not have a 'time' dimension.
+    """
+    if not any(isinstance(data, dtype) for dtype in (xr.DataArray, xr.Dataset)):
+        raise ValueError(
+            "Input data has to be an xarray-DataArray or xarray-Dataset, "
+            f"not {type(data)}"
+        )
+    if "time" not in data.dims:
+        raise ValueError(
+            "Analysis is done of the 'time' dimension, but the input data"
+            f" only has dims: {data.dims}"
+        )
+
+
+class Preprocessor:
+    """Preprocessor for s2s data."""
```

Comment:
It is not really only for 's2s' data, right? More generally for climate data?

Reply:
Or to be even more general, spatial-temporal data.

Reply:
Corrected!